### ISSUE TITLE
Windows version?

### ISSUE THREAD

**Issue by mikera** [uncomplicate/neanderthal-atlas#2]
I was trying to use neanderthal-atlas on Windows but couldn't get it to work (fails with "java.lang.RuntimeException: Library 'neanderthal-atlas-0.1.0.dll' not found!").

Should neanderthal-atlas work on Windows? Or have I just configured something wrong?

**Comment by blueberry** [comment]
Hi Mike,

The library should work on windows (there is nothing OS specific in it). I haven't built it for Windows (yet), though. The build is automatic, and mavenized! so you can try to build it yourself. Please see Issue #1 for what kind of changes might be necessary (if any). Please send me the build if/when you build it, so I can upload it to clojars.
Also worth reading is installation instructions at neanderthal.uncomplicate.org

**Comment by mikera** [comment]
Thanks for the pointers!

I've tried the Maven build but it seems to be failing due to the (obvious) lack of `gcc`. I'm not familiar with the `nar-maven-plugin`, but not even sure if even the presence of `gcc` would be helpful though if the aim is to produce a `.dll`? Is there some sort of extra configuration required?

```
[INFO]
[INFO] --- nar-maven-plugin:3.2.0:nar-validate (default-nar-validate) @ neandert
hal-atlas ---
[INFO] Using AOL: amd64-Windows-gpp
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6.396 s
[INFO] Finished at: 2015-06-30T18:02:10+01:00
[INFO] Final Memory: 14M/358M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.2.0:nar-v
alidate (default-nar-validate) on project neanderthal-atlas: Could not launch cm
d.exe /X /C "gcc --version": Error while executing process. Cannot run program "
gcc": CreateProcess error=2, The system cannot find the file specified -> [Help
1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e swit
ch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please rea
d the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionE
xception
```

**Comment by blueberry** [comment]
I don't have Windows, so I do not know for sure, but what I am sure is that you can configure lots of options in NAR, one of them is the compiler (you'll have to check the details at NAR's web site).

Additionally, the build will expect that atlas has been built and configured on your system. As ATLAS is *nix oriented, this will almost definitely require installing MinGW, which includes gcc, which is needed for building ATLAS anyway. Now, you can skip that step and use pre-built atlas for Windows, but I do not know what will be the performance penalty later.

neanderthal's CBLAS layer might even be compiled with other CBLAS compatible libraries (OpenBLAS for example) that might be a better match for Windows, but unfortunately I am not able to experiment with that because I do not have a Windows machine (but I will definitively nail that problem once neanderthal approaches maturity and the functional details are polished).

Yes, I know this stuff is cumbersome, but that's how it goes with native stuff... Fortunately, once it is built there is a way to be included in Java without so much fuss.

**Comment by mikera** [comment]
Hmmm tricky. Getting cross platform support for native code is always such a nuisance. 

Did you ever look at netlib-java?
https://github.com/fommil/netlib-java

Full disclosure: I'm looking at this because I want to figure out a good way of getting optimised native linear algebra code working with `core.matrix`. `neanderthal-atlas` might be an option, but only if the cross platform support can be made to work. Improving `Clatrix` is also an option.

**Comment by blueberry** [comment]
I've done my homework and looked at them all. The problem with all the options was: when you want to avoid the nuisance and use the default, you get a (sometimes much) slower version of the library, and also have to deal with lots of poorly documented stuff (if there is any documentation at all). If there is a Java alternative that is just a bit slower, why bother? If you want to be sure that you'll get the (worthy) full performance, you'll have to deal with the dirty stuff even with those libraries.

Now, that dirty stuff can be dealt through that Java library, or, (in my opinion a better option) be left to the independent platform installation. In that way, the Java library can be relatively simple, easy and straightforward, while the responsibility of the developer that is using it is to know his platform and install the native library. Yes, he still has to do the work, but it is much more likely that he will find a decent information on how to do it with vanilla native library than with some obscure Java integration layer that tries to solve the problem that remains poorly solved for decades.

And, the final advice: benchmark whatever you choose...

Just my 2 euro cents, I am curious to see what you'll come up with ;)

**Comment by mikera** [comment]
Understand what you are saying. It is certainly true that if you want to squeeze out the last drop of performance then you need to really know your platform, and compile native libraries accordingly.

However my experience in general purpose situations is that if the install experience is too difficult then a lot of people simply won't bother. Not many developers (sadly? or maybe just reality?) have the time to research and create their own platform specific builds. So it becomes the responsibility of us who create libraries to find a way to make this work (at least, if we want the libraries to be useful and widely adopted).

Ultimately performance is a secondary concern 99% of the time: in the real world, practicalities of data integration, getting a smooth CI build process, getting the data in the right format etc. are almost always more important than raw performance. I personally don't spend much time on optimisation unless I actually hit a bottleneck (which for me is mostly around 1D vector operations in vectorz-clj.... If you want to benchmark Neanderthal against vectorz-clj with those, then I'll happily take up the challenge :-) ).

Having said that, I'd like to ensure that the optimal performance is _achievable_ for those who really need it.

That's the beauty of `core.matrix` by the way: you can code to a standard, idiomatic Clojure API, but switch to better implementations based on your performance or platform requirements (which might include a platform-specific optimised build). Fairly big code bases (including e.g. Incanter) have proved that this is possible with minimal overhead.

So I guess what I'm looking for is:
a) A fast native matrix library (i.e. significantly better than Clatrix which is a bit of a kludge)
b) A relatively painless cross-platform experience for users and developers (at least by default, if users want to do extra work to get faster performance for their specific platform that is fine...)
c) Proper integration with core.matrix

Looks like you have done a) with `neanderthal-atlas`, I'm certainly happy to do c), but b) needs to work  otherwise the effort is IMHO a bit pointless :-)

FWIW here is my experimental repo trying to integrate `neanderthal-atlas` into `vectorz-clj`, which would automatically fix c). I think it should work, but I've just hit a bit of a roadblock getting the thing to build on Windows :-)
- https://github.com/mikera/vectorz-native

**Comment by blueberry** [comment]
b) is solved for windows in the upcoming release 0.8.0 of neanderthal-native.

**Comment by mikera** [comment]
Thanks for taking a look at this. I will test out 0.8.0 when it is released and contribute c) If I can get it working.
